lib/Dialect/Transforms/FusionPatterns.cpp

```diff
@@ -12,70 +12,145 @@
 #include "mlir-tcp/Dialect/IR/TcpOps.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
 #include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dominance.h"
 #include "mlir/IR/OpDefinition.h"
+#include "llvm/Support/Debug.h"
+
+#ifndef NDEBUG
+#define DEBUG_TYPE "tcp-fusion-patterns"
+#endif
 
 namespace mlir::tcp {
+
 LogicalResult
 GenericBottomUpFuser::matchAndRewrite(Operation *op,
                                       PatternRewriter &rewriter) const {
-  Operation *use = op;
-  bool isChanged = false;
-  for (auto operand : op->getOperands()) {
-    if (operand.getDefiningOp()) {
-      Operation *def = operand.getDefiningOp();
-      if (canFuse(def, use)) {
 
-        // Currently we are only fusing ops at the top-level.
-        // This is to avoid recursing inside a group and ending up with
-        // nested groups that contain the same ops.
-        // Since we are iterating bottom up in a block, we only need to check
-        // if the def op has a func parent.
-        //
-        // TODO: Remove this restriction to allow fusing in nested regions.
-        if (!isa<func::FuncOp>(def->getParentOp())) {
-          continue;
-        }
+  // Currently we are only fusing ops at the top-level.
+  // This is to avoid recursing inside a group and ending up with
+  // nested groups that contain the same ops.
+  // Since we are iterating bottom up in a block, we only need to check
+  // if the def op has a func parent.
+  //
+  // TODO: Remove this restriction to allow fusing in nested regions.
+  if (!isa<func::FuncOp>(op->getParentOp())) {
+    return failure();
+  }
```

Comment:
style nit: prefer omitting braces for single-statement ifs ([reference](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements))